Feature/get dotted columns by prefix

BibMartin opened the pull request:
- [ ] closes #xxxx
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [ ] whatsnew entry

Assuming a DataFrame like this:
```
>> df
    foo  bar.foo  bar.bar
0  0.32     0.15     0.45
1  0.62     0.73     0.36
2  0.21     0.68     0.10
3  0.05     0.36     0.90
```

one can access to `df['foo']` with the shortcut `df.foo`.

This PR extends this idea to column names containing dots, so that `df.bar` yields:

```
>> df.bar
    foo   bar
0  0.15  0.45
1  0.73  0.36
2  0.68  0.10
3  0.36  0.90
```

I'm new to contributing in pandas: so I need a few help:

* Do I need to create an issue to *host* the PR ?
* Where in the docs should I host explainations about this ?

Thanks in advance (and pandas is awesome btw !)

Commit:
ENH: Access sub-dataframe with the same dot-separated-prefix

Commit:
TST: test attibute indexing on dotted-prefixed-columns

TomAugspurger commented:
> Do I need to create an issue to host the PR ?

Not necessary, but typically we use issues to talk about whether a change is likely to be accepted, so that a contributor doesn't spend time on something that doesn't fit well in pandas, which might be relevant here :)

I'll let others chime in, but this might not receive support from the other maintainers. My initial reaction is -1 since it complicates when `.` access will work, and what will be returned if it does. Right now, it works when

- the column name is a valid python identifier
- the column name doesn't shadow an existing method

and always returns the same thing as `df[column_name]`.

I think this change would complicate that, `df[prefix]` is a KeyError, but `df.prefix` would potentially return some columns.

chris-b1 commented:
This is an interesting idea, but I'm slightly negative on this being accepted too.  I think a `MultiIndex` already handles this usecase reasonable well?

```
def make_col(s):
    pieces = s.split('.')
    if len(pieces) == 1:
        return tuple([pieces[0], ''])
    else:
        return tuple(pieces)

df.columns = pd.MultiIndex.from_tuples(df.columns.map(make_col))

df
Out[61]: 
    foo   bar      
          foo   bar
0  0.32  0.15  0.45
1  0.62  0.73  0.36
2  0.21  0.68  0.10
3  0.05  0.36  0.90

df['bar']
Out[62]: 
    foo   bar
0  0.15  0.45
1  0.73  0.36
2  0.68  0.10
3  0.36  0.90

df.bar
Out[63]: 
    foo   bar
0  0.15  0.45
1  0.73  0.36
2  0.68  0.10
3  0.36  0.90
```

BibMartin commented:
@chris-b1 
>  I think a MultiIndex already handles this usecase reasonably well?

That's true. Your code works when you know the number of levels to be created. But this can handle the general case with few effort:

```
nb_level = df.columns.map(lambda x: len(x.split('.'))).max()
df.columns = df.columns.map(lambda x: tuple((x.split('.') + ['']*nb_level)[:nb_level]))
```

This being said, plus @TomAugspurger's remark on `df.prefix` being different from `df['prefix']`, I guess this PR can be closed.

Thanks for the review and comments.
